Source/GSOrderedSet.m

```diff
@@ -0,0 +1,381 @@
+}
+
+- (NSUInteger) countByEnumeratingWithState: (NSFastEnumerationState*)state
+                                   objects: (id*)stackbuf
+                                     count: (NSUInteger)len
+{
+  //state->mutationsPtr = (unsigned long *)self;
+  //return GSIMapCountByEnumeratingWithStateObjectsCount
+  //            (&map, state, stackbuf, len);
+
+  return [self count];
+}
+
+- (NSUInteger) sizeInBytesExcluding: (NSHashTable*)exclude
+{
+  NSUInteger	size = GSPrivateMemorySize(self, exclude);
+
+  if (size > 0)
+    {
+      NSUInteger count = [self count];
+      NSUInteger i = 0;
+
+      for(i = 0; i < count; i++)
+	{
+	  GSIArrayItem item = GSIArrayItemAtIndex(&array, i);
+          size += [item.obj sizeInBytesExcluding: exclude];
+        }
+    }
+  return size;
+}
+
+// Put required overrides here...
+- (NSUInteger) count
+{
+  return GSIArrayCount(&array);
+}
+
+- (id) objectAtIndex: (NSUInteger)index
+{
+  GSIArrayItem item = GSIArrayItemAtIndex(&array, index);
+  return item.obj;
+}
+
+/* Designated initialiser */
+- (id) initWithObjects: (const id*)objs count: (NSUInteger)c
+{
+  NSUInteger i = 0;
+
+  // Initialize and fill the set.
+  GSIArrayInitWithZoneAndCapacity(&array, [self zone], c);
+  for (i = 0; i < c; i++)
+    {
+      id obj = objs[i];
+      GSIArrayItem item;
+      
+      if (objs[i] == nil)
+	{
+	  DESTROY(self);
+	  [NSException raise: NSInvalidArgumentException
+		      format: @"Tried to init set with nil value"];
+	}
+
+      item.obj = obj;
+      if(![self containsObject: obj])
+	{
+	  GSIArrayAddItem(&array, item);
+	  RETAIN(obj);
+	}
+    }
+  return self;
+}
+
+@end
+
+@implementation GSMutableOrderedSet
+
++ (void) initialize
+{
+  if (self == [GSMutableOrderedSet class])
+    {
+      GSObjCAddClassBehavior(self, [GSOrderedSet class]);
+    }
+}
+
+- (void) addObject: (id)anObject
+{
+  GSIArrayItem item;
+
+  if (anObject == nil)
+    {
+      [NSException raise: NSInvalidArgumentException
+		  format: @"Tried to add nil to set"];
+    }
+
+  if([self containsObject: anObject] == NO)
+    {
+      item.obj = anObject;
+      GSIArrayAddItem(&array, item);
+      RETAIN(anObject);
+      _version++;
+    }
+}
+
+- (void) insertObject: (id)object atIndex: (NSUInteger)index
+{
+  GSIArrayItem item;
+  if([self containsObject: object] == NO)
+    {
+      item.obj = object;
+      GSIArrayInsertItem(&array, item, index);
+      RETAIN(object);
+      _version++;
+    }
+}
+
+- (void) _insertObject: (id)object atIndex: (NSUInteger)index
+{
+  GSIArrayItem item;
+  
+  item.obj = object;
+  GSIArrayInsertItem(&array, item, index);
+  RETAIN(object);
+  _version++;
+}
+
+- (void) removeObjectAtIndex: (NSUInteger)index
+{
+  _version++;
+  GSIArrayRemoveItemAtIndex(&array, index);
+}
+
+- (void) replaceObjectAtIndex: (NSUInteger)index
+		   withObject: (id)obj
+{
+  [self _insertObject: obj atIndex: index];
+  [self removeObjectAtIndex: index];
```

Comment:
Shouldn't you remove the old object first? This looks like you remove the newly added object. Did you write a test for this case?